### ISSUE TITLE
linux compatibility attempt progress

### ISSUE THREAD

**Issue by TheK0tYaRa** [cheat-engine/cheat-engine#1963]
I think at this point it's worth it to make a single issue for all issues i have while making it work on linux.
The issues i've had before: #1959 #1962 
Gonna ask if issues arrive here next.

**Comment by TheK0tYaRa** [comment]
I suppose that `dwProcessId` and `dwThreadId` are `pid_t` and `pthread_t`

**Comment by TheK0tYaRa** [comment]
![image](https://user-images.githubusercontent.com/27967034/147737140-8bb54edb-bae3-4bc8-84ff-475dce166c11.png)
So a type block can't read variables. Excellent. Preseeding go brrrr

**Comment by cheat-engine** [comment]
cobst has to go first

**Comment by TheK0tYaRa** [comment]
nvm, it was implemented elsewhere already.

**Comment by TheK0tYaRa** [comment]
Can't add unit `classesh.inc` to project because with it it adds `<IsPartOfProject Value="True"/>` to the unit. Not sure how it changes anything but still.

**Comment by TheK0tYaRa** [comment]
I have made all the changes without making commits to a local branch so far. I won't continue doing so further. Partially because i have got this issue currently and partially because i've realised that it was dumb to not make snapshots.
`environmentopts.pp(52,3) Fatal: Cannot find DbgIntfDebuggerBase used by EnvironmentOpts, incompatible ppu=/usr/lib/lazarus/components/debuggerintf/lib/x86_64-linux/dbgintfdebuggerbase.ppu, package DebuggerIntf`

**Comment by TheK0tYaRa** [comment]
I currently can not continue doing stuff without guidance on that issue.

**Comment by cheat-engine** [comment]
go to project->inspector and remove everything related to debugger and what else you may have added yourself

Also create a new project and then make sure you can at least compile applications, in case your lazarus config is broken

**Comment by TheK0tYaRa** [comment]
![image](https://user-images.githubusercontent.com/27967034/148816290-2d5ddaff-7831-4630-9e55-824372d11678.png)
That's the exact component name

**Comment by TheK0tYaRa** [comment]
![image](https://user-images.githubusercontent.com/27967034/148816841-c6d1ca0d-a275-4021-8acf-055fa47cf565.png)
removing all the mentions of `debugger changed nothing`

**Comment by TheK0tYaRa** [comment]
And the project that i've so far used for testing the compatibility stuff still compiles fine

**Comment by cheat-engine** [comment]
Restore the lpi and lpr with the original and see if it still happens.

**Comment by TheK0tYaRa** [comment]
restored, created a new target from win64, removed `-WB00400000` that it complained about, compiled, same error.

**Comment by cheat-engine** [comment]
target os is linux right ?

Anyhow, if nothing else works, just create a whole new project and add all the units one by one back to your project.
Also, make sure not a single unit adds the windows or jwawindows unit in the uses clause  (look at how the darwin define deals with that)

**Comment by TheK0tYaRa** [comment]
Yeah, there is a ton of ifdefs i've made so far xD
I'll let you know if it breaks at that point

**Comment by TheK0tYaRa** [comment]
Well, that's confusing. This error didn't occur on the original project.
`frmstructuresconfigunit.pas(460,4) Fatal: Cannot open include file "frmstructuresconfigunit.lrs"`

**Comment by TheK0tYaRa** [comment]
And in the original project the said "missing" file is automatically generated

**Comment by TheK0tYaRa** [comment]
So currently stuck on `formdesignerunit.pas(202,57) Fatal: Cannot find LazMsgDialogs used by formdesignerunit.` , Haven't found it in any of "traditional" folders yet, FPC and lazarus folders included.

**Comment by cheat-engine** [comment]
just strip that unit and code for now and remove all references to it

**Comment by TheK0tYaRa** [comment]
Noted. The main reason why i am using a newer lazarus is specifics of my linux distro (it can't downgrade or it dies, archlinux biggest weakness)

**Comment by TheK0tYaRa** [comment]
![image](https://user-images.githubusercontent.com/27967034/150043822-b046d8bb-a773-4d3c-a42b-5be9de0e1a3f.png)
Finally - empty space gave me an exception xD

**Comment by cheat-engine** [comment]
Do run->build, not  run->compile

**Comment by TheK0tYaRa** [comment]
Nah, it's not that. I just recently bought RAM. I need to test it for defects.
Also doing the Shift F9 doesn't change the Messages output - it's still randomly an exception.

**Comment by TheK0tYaRa** [comment]
`Error: Duplicate resource: Type = 24, Name = 1, Lang ID = 0000`
That sounds like something almost ready to run. But i dunno why it recommends adding -vd but it doesn't add any output. Maybe i'm filtering them out, one minute

**Comment by cheat-engine** [comment]
remove the {$ lines from the .lpr} and see how that goes

**Comment by TheK0tYaRa** [comment]
Doing the search by resource name within the project and within fpc folder gave no result, brute comment spam time!

**Comment by TheK0tYaRa** [comment]
Apparently i'm too lazy to try and check which one are captured by `{$R *.res}`

**Comment by TheK0tYaRa** [comment]
The current issue is
`Error: lua.pas:(.text.n_LUA_$$_LUA_GETFENV$POINTER$LONGINT+0x1): undefined reference to `lua_getfenv'`

**Comment by TheK0tYaRa** [comment]
So it's lua

**Comment by TheK0tYaRa** [comment]
I think that the library has compatibility issues, as even lua53 has issues, being 2 years old after last update.

**Comment by TheK0tYaRa** [comment]
Imma skip all lua stuff for now

**Comment by TheK0tYaRa** [comment]
![image](https://user-images.githubusercontent.com/27967034/150471246-09f34d10-c981-4209-b253-b1bc0738377d.png)
Finally, time to fix parts now. The reason for previous error was that some lua functions are no more.

**Comment by TheK0tYaRa** [comment]
Got an issue with LuaHandler lines 1374 and 1376 throwing exceptions. That issue i have no idea how to fix yet. Except maybe implementing lua functions that are missing.

**Comment by TheK0tYaRa** [comment]
It's an undefined lua error to be more exact. I suppose attaching a screenshot might help the issue.
![image](https://user-images.githubusercontent.com/27967034/150632151-430c8d2b-7e77-440d-afcd-5e76509549e8.png)

**Comment by TheK0tYaRa** [comment]
what's the `_LuaCS.Enter` doing and how do i smack it to make it work, hmm

**Comment by TheK0tYaRa** [comment]
oh, right, `{ #todo : not in lua < 5.1 }//    luaL_openlibs(L);`

**Comment by TheK0tYaRa** [comment]
`Exception=Access violation` cool, time to start running a debug build as root

**Comment by cheat-engine** [comment]
do a stacktrace.  Likely a lua export not found.

Also, use the novmx define when debugging.  You likely don't have dbvm running

**Comment by TheK0tYaRa** [comment]
As a madlad who debugged before today with a console log spam, i have only learned today what is a breakpoint **in code**. I'm gonna suppose that you meant the next by a stack trace as i haven't learned what it looks like in pascal.
```
MainUnit2
[FORMS.PP] ExceptionOccurred 
  Sender=EAccessViolation
  Exception=Access violation
  Stack trace:
  $0000000000000000
[FORMS.PP] ExceptionOccurred 
An unhandled exception occurred at $00007FCE0182D745:
EAccessViolation: Access violation
  $00007FCE0182D745
```

**Comment by TheK0tYaRa** [comment]
But none the less i know at which line (more or less) the exception happens at - `ownprocesshandle := OpenProcess(ifthen(GetSystemType<=6,$1f0fff, process_all_access), True, GetCurrentProcessId);` aka line 3837 in CEFuncProc

**Comment by TheK0tYaRa** [comment]
oh, GetCurrentProcessId is not yet written, smh

**Comment by TheK0tYaRa** [comment]
![image](https://user-images.githubusercontent.com/27967034/150638809-df660bc5-c9ec-4646-b623-c6701b6ac84c.png)
huh... Is that a type conversion error or...?

**Comment by TheK0tYaRa** [comment]
Got an issue finding out an implementation for OpenProcess as it's a winapi piece.

**Comment by cheat-engine** [comment]
Emulate it

Create your own handle table and use that to link it with a certain process

look at ceserver api.c. it does that as well

**Comment by TheK0tYaRa** [comment]
How to convert C `TAILQ_ENTRY()` to a type?

**Comment by cheat-engine** [comment]
Use a TQueue from the Contnrs uses instead

**Comment by TheK0tYaRa** [comment]
Let me guess - `TAILQ_HEAD()` is going to be TObjectQueue ?
